modules/saslplainauth.cpp

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2004-2025 ZNC, see the NOTICE file for details.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <znc/User.h>
+#include <znc/znc.h>
+
+class CSASLMechanismPlain : public CModule {
+  public:
+    MODCONSTRUCTOR(CSASLMechanismPlain) { AddHelpCommand(); }
+
+    void OnClientGetSASLMechanisms(SCString& ssMechanisms) override {
+        ssMechanisms.insert("PLAIN");
+    }
+
+    EModRet OnClientSASLAuthenticate(const CString& sMechanism,
+                                     const CString& sMessage) override {
+        if (!sMechanism.Equals("PLAIN")) {
+            return CONTINUE;
+        }
+
+        CString sNullSeparator = std::string("\0", 1);
+        CString sAuthzId = sMessage.Token(0, false, sNullSeparator, true);
+        CString sAuthcId = sMessage.Token(1, false, sNullSeparator, true);
+        CString sPassword = sMessage.Token(2, false, sNullSeparator, true);
```

Comment:
Having no clue about sasl.... what is up with the 3rd or 4th field? Should the last call use `true` instead of `false` so that all of them are just part of the password?

Reply:
From [RFC 4616](https://datatracker.ietf.org/doc/html/rfc4616)

```
   message   = [authzid] UTF8NUL authcid UTF8NUL passwd
   passwd    = 1*SAFE ; MUST accept up to 255 octets
   UTF8NUL   = %x00 ; UTF-8 encoded NUL character

   SAFE      = UTF1 / UTF2 / UTF3 / UTF4
               ;; any UTF-8 encoded Unicode character except NUL
```

Because 4th field doesn't exist, and password cannot contain NUL itself, we can either ignore the extra fields, or reject auth if they are present

Reply:
That said, we probably should replace the extensive usage of Token() throughout the codebase with some kind of Split(), to avoid going through first fields multiple times. And replace many usages of CString with std::string_view. But that's outside of scope of this PR